grouper/background.py

```diff
@@ -67,12 +75,51 @@ def expire_edges(self, session):
             edge.active = False
             session.commit()
 
+    def expire_nonauditors(self, session):
+        """Checks all enabled audited groups and ensures that all approvers for that group have
+        the PERMISSION_AUDITOR permission. All approvers of audited groups that aren't auditors
+        have their membership in the audited group set to expire
+        settings.nonauditor_expiration_days days in the future.
+
+        Args:
+            session (Session): database session
+        """
+        now = datetime.utcnow()
+        graph = Graph()
+        exp_days = timedelta(days=settings.nonauditor_expiration_days)
+        # Hack to ensure the graph is loaded before we access it
+        graph.update_from_db(session)
+        for group in get_audited_groups(session):  # TODO(tyleromeara): replace with graph call
+            members = group.my_members()
+            for (type_, member), edge in members.iteritems():
+                # Auditing is already inherited, so we don't need to handle that here
+                if type_ == "Group":
+                    continue
+                member = User.get(session, name=member)
+                if user_role_index(member, members) not in APPROVER_ROLE_INDICIES:
+                    continue
+                if user_has_permission(session, member, PERMISSION_AUDITOR):
+                    continue
+                edge = GroupEdge.get(session, id=edge.edge_id)
+                if edge.expiration and edge.expiration < now + exp_days:
+                    continue
+                exp = now + exp_days
+                exp = exp.date()
+                edge.apply_changes_dict(
+                    {"expiration": "{}/{}/{}".format(exp.month, exp.day, exp.year)}
+                )
+                edge.add(session)
+                notify_nonauditor_flagged(settings, session, edge)
+        session.commit()
+
     def run(self):
         while True:
             try:
                 session = Session()
                 logging.debug("Expiring edges....")
                 self.expire_edges(session)
+                logging.debug("Expiring nonauditor approvers in audited groups...")
```

Comment:
How slow is this? Would this make background thread runtime *significantly* longer on a Dropbox-sized Grouper install?